src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs

```diff
@@ -235,8 +237,27 @@ public static async Task<PackageSpec> GetProjectSpec(IDependencyGraphProject pro
                     {
                         dgSpec.AddRestore(packageSpec.RestoreMetadata.ProjectUniqueName);
                     }
+
+                    var projFileName = Path.GetFileName(packageSpec.RestoreMetadata.ProjectPath);
+                    var dgFileName = string.Format(DependencyGraphSpec.DGSpecFileName, projFileName);
+                    var persistedDGSpecPath = Path.Combine(packageSpec.RestoreMetadata.OutputPath, dgFileName);
+
+                    if (File.Exists(persistedDGSpecPath))
+                    {
+                        var persistedDGSpec = DependencyGraphSpec.Load(persistedDGSpecPath);
+
+                        foreach (var dependentPackageSpec in persistedDGSpec.GetClosure(packageSpec.RestoreMetadata.ProjectUniqueName))
+                        {
+                            if (!projects.Any(p => PathUtility.GetStringComparerBasedOnOS().Equals(p.MSBuildProjectPath, dependentPackageSpec.RestoreMetadata.ProjectPath)) &&
```

Comment:
1.  If you have `n` projects and all projects reference the same unloaded project, we'll perform this check up to `n^2` times.  I would add an initial check:

         !uniqueProjectDependencies.Contains(dependentPackageSpec.RestoreMetadata.ProjectUniqueName) &&

2.  Outside these loops (like [here](https://github.com/NuGet/NuGet.Client/pull/2521/files#diff-66b6645a4772b600680f93c7fee07f5aR220) you can memoize `PathUtility.GetStringComparerBasedOnOS()` with a local and use it here and [here](https://github.com/NuGet/NuGet.Client/pull/2521/files#diff-66b6645a4772b600680f93c7fee07f5aR221).